ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncSocketImpl.kt

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2018-2020 NetFoundry, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.netfoundry.ziti.net.nio
+
+import java.io.InputStream
+import java.io.OutputStream
+import java.net.*
+import java.nio.ByteBuffer
+import java.nio.channels.AsynchronousSocketChannel
+import java.nio.channels.CompletionHandler
+import java.util.concurrent.*
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.math.min
+
+/**
+ * SocketImpl wrapper over AsyncSocketChannel
+ */
+internal class AsyncSocketImpl(ch: AsynchronousSocketChannel? = null): SocketImpl() {
+    internal lateinit var channel: AsynchronousSocketChannel
+
+    internal val inputLock = Semaphore(1)
+    internal val input = ByteBuffer.allocate(32 * 1024).apply {
+        flip()
+    }
+    internal val timeout = AtomicInteger(Socket().soTimeout)
+
+    init {
+        ch?.let {
+            channel = it
+        }
+    }
+
+    override fun bind(host: InetAddress?, port: Int) = error("only client sockets are supported")
+    override fun accept(s: SocketImpl?) = error("only client sockets are supported")
+    override fun listen(backlog: Int) = error("only client sockets are supported")
+
+    override fun getInetAddress() = (channel.remoteAddress as InetSocketAddress?)?.address
+    override fun getPort() = (channel.remoteAddress as InetSocketAddress?)?.port ?: 0
+    override fun getLocalPort() = (channel.localAddress as InetSocketAddress?)?.port ?: 0
+
+    override fun shutdownInput() {
+        channel.shutdownInput()
+    }
+
+    override fun shutdownOutput() {
+        channel.shutdownOutput()
+    }
+
+    override fun create(stream: Boolean) {
+        require(stream){"only stream sockets are supported"}
+        if (!::channel.isInitialized) {
+            channel = AsynchronousSocketChannel.open()
+        }
+    }
+
+    override fun connect(host: String, port: Int) =
+        connect(InetSocketAddress(host, port), 0)
+
+    override fun connect(address: InetAddress, port: Int) = connect(
+        InetSocketAddress(address, port), 0)
+
+    override fun connect(address: SocketAddress, timeout: Int) {
+        val cf = channel.connect(address)
+        if (timeout > 0) {
```

Comment:
this should have a floor associated with it right?

Reply:
timeout is checked in the base Socket before impl is called